CMakeLists.txt

```diff
@@ -12,38 +10,52 @@ set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake)
 
 # Make sure that CMAKE_BUILD_TYPE is either Debug or Release:
 if (NOT CMAKE_BUILD_TYPE)
-    set(CMAKE_BUILD_TYPE Release
+    set(CMAKE_BUILD_TYPE RelWithDebInfo
         CACHE STRING "Build type (Debug, Release)" FORCE)
 endif ()
 if (NOT (CMAKE_BUILD_TYPE STREQUAL "Debug" OR
-        CMAKE_BUILD_TYPE STREQUAL "Release"))
-    message("${CMAKE_BUILD_TYPE}")
-    message(FATAL_ERROR "CMAKE_BUILD_TYPE must be one of: Debug, Release (current value: '${CMAKE_BUILD_TYPE}')")
+        CMAKE_BUILD_TYPE STREQUAL "Release" OR
+        CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo" OR
+        CMAKE_BUILD_TYPE STREQUAL "MinSizeRel"))
+    message(FATAL_ERROR "CMAKE_BUILD_TYPE must be one of: Debug, Release, RelWithDebInfo, MinSizeRel (current value: '${CMAKE_BUILD_TYPE}')")
+endif ()
+
+if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    # g++
+    set(CMAKE_CXX_FLAGS "-std=c++11 -Wall -Wextra -fno-common ${CMAKE_CXX_FLAGS} ")
+    if (MINGW)
+        set(CMAKE_CXX_FLAGS "-D_hypot=hypot -Doff64_t=_off64_t ${CMAKE_CXX_FLAGS} ")
+        if("${CMAKE_SIZEOF_VOID_P}" STREQUAL "8")
+            set(CMAKE_CXX_FLAGS "-DMS_WIN64 ${CMAKE_CXX_FLAGS}")
+        endif()
+    endif()
+elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Intel")
+    # icpc
+    set(CMAKE_CXX_FLAGS "-std=c++11 -Wall -fno-common ${CMAKE_CXX_FLAGS}")
+elseif (CMAKE_CXX_COMPILER_ID MATCHES Clang)
+    # clang
+    set(CMAKE_CXX_FLAGS "-std=c++11 -Wall -Wextra ${CMAKE_CXX_FLAGS}")
+elseif (CMAKE_CXX_COMPILER_ID STREQUAL "PGI")
+    # pgcpp
+    set(CMAKE_CXX_FLAGS "--gnu --c++11 -D__GXX_EXPERIMENTAL_CXX0X__ ${CMAKE_CXX_FLAGS}")
 endif ()
 
-set(BUILD_FOR_DISTRIBUTION no
-    CACHE BOOL "Building SymEngine for a distribution")
+set(CMAKE_POSITION_INDEPENDENT_CODE yes CACHE BOOL "Build all libraries as position independent")
 
 #Boost.multiprecision build segfaults if -O3 optimization is on
 #which it is by default for cmake Release mode. Remove -O3 from flags.
 if (("${INTEGER_CLASS}" MATCHES "boostmp") AND CMAKE_CXX_COMPILER_ID MATCHES "Clang")
     string(REPLACE "-O3" "" CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}")
+    string(REPLACE "-O2" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
 endif()
 
 if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
-    set(CMAKE_CXX_FLAGS_DEBUG
-        "${CMAKE_CXX_FLAGS_DEBUG} -Wno-unused-parameter -ggdb")
-    set(CMAKE_CXX_FLAGS_RELEASE
-        "${CMAKE_CXX_FLAGS_RELEASE} -Wno-unused-parameter")
-    if (NOT BUILD_FOR_DISTRIBUTION)
-        set(CMAKE_CXX_FLAGS_RELEASE
-            "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
```

Comment:
I've removed this and then checked the benchmarks and there were no noticeable difference. @certik, can you check?

Reply:
The `march=native` makes a big difference on some computers. I would leave it there for now.